### PR TITLE
0.5.0-rc15: Optional properties have now 50% chances to be generated.

### PR DESCRIPTION
**Pull request by manuelmhtr**
## Problem
Currently, the library generates only up to 5 optional properties (usually only 1 - 3). This number does not consider schema's size (ie. 3 optional properties might be a lot for small schemas, but almost nothing for big ones).

## Solution
Instead of a fixed limit, the number of optional properties being generated is a random number between required properties and total properties.

That means, each optional property has 50% chances to be generated.

### CODE REVIEW

**Comment by pateketrueke** [comment]
Hi, thanks for your contribution!

Definitely this would work but please just make the changes on the `ts` sources.

Also, to make a proper release for this you must run the `build` tasks to generate the final sources.

- Rebase your current work from the `develop` branch
- Change the `version` from package.json
- Run `npm run dist`

I'm not sure if this should be back-ported to `0.4.x` but I think it's better focusing on releasing a stable `0.5.x` soon. 🍻

**Comment by pateketrueke** [comment]
Hi, I've been playing with your changes but they were breaking most tests...

On this case, I just leave `shuffle(patternPropertyKeys)` as it was because it's necessary for taking into account `patternProperties` too.

I clearly see the value of this change, but we need to ensure we're doing it right... how we can test that it'll work as expected?

I didn't have a chance to actually think about it, but if you have some examples of large schemas to run against it can help us to test this out.

Thanks for your valuable input and feedback.

I'm glad this is helping you.

File: spec/schema/core/types/object.json [hunk 1]
```diff
@@ -2,81 +2,81 @@
...
     "description": "object generator",
     "tests": [
-        {
-          "description": "addP === false && !props && !patternProps",
-          "schema": {
-            "type": "object",
-            "additionalProperties": false
-          },
-          "valid": true
+      {
```
Comment (manuelmhtr) [inline spec/schema/core/types/object.json]: Tests are the same, just fixed indentation.

File: ts/types/object.ts [hunk 2]
```diff
@@ -18,6 +18,11 @@ var objectType: FTypeGenerator = function objectType(value: IObjectSchema, path,
 
     var propertyKeys = Object.keys(properties);
     var patternPropertyKeys = Object.keys(patternProperties);
+    var allProperties = propertyKeys.concat(patternPropertyKeys);
```
Comment (manuelmhtr) [inline ts/types/object.ts]: My bad, I was ignoring `patternPropertyKeys`.

File: ts/types/object.ts [hunk 3]
```diff
@@ -43,22 +48,18 @@ var objectType: FTypeGenerator = function objectType(value: IObjectSchema, path,
...
-
     var min = Math.max(value.minProperties || 0, requiredProperties.length);
-    var max = Math.max(value.maxProperties || random.number(min, propertyKeys.length));
-
-    random.shuffle(patternPropertyKeys).forEach(function(_key) {
-        if (requiredProperties.indexOf(_key) === -1) {
-            requiredProperties.push(_key);
-        }
-    });
+    var max = Math.max(value.maxProperties || random.number(min, allProperties.length));
```
Comment (manuelmhtr) [inline ts/types/object.ts]: This now includes `propertyKeys` and `patternPropertyKeys`.

File: spec/schema/core/option/random.js [hunk 4]
```diff
@@ -3,7 +3,7 @@ const seedrandom = require('seedrandom');
 module.exports = {
     register: function(jsf) {
         return jsf.option({
-            random: seedrandom('some seed')
+            random: seedrandom('some seed', {state: true})
```
Comment (pateketrueke) [inline spec/schema/core/option/random.js]: Why you added `state: true`?

I just checked a reference for `seedrandom` but still I don't get it.

**Comment by manuelmhtr** [comment]
Closing PR because this approach doesn't work with schema described in [issue 336](https://github.com/json-schema-faker/json-schema-faker/issues/336).

Working in an alternative solution.

**Pull request closed without merging**